Adding youtuber blacklist feature

JeroenDeDauw opened the pull request:
Based on https://github.com/FAForever/fafbot/pull/22, **which should be merged and reviewed first**

JeroenDeDauw commented:
Youtubers can now be blacklisted in the config file with the new `blacklisted_youtubers` option.

Note that the implementation for this is simplistic. It just fetches twice the needed casts and then filters. So it is possible, though quite unlikely, that less than 5 results will be shown.

JeroenDeDauw commented on src/MessageHandler.py:
```diff
@@ -0,0 +1,40 @@
+from src.StreamsCommandHandler import StreamsCommandHandler
+
+
+class MessageHandler:
```
I could stuff this code all into fafbot.py yes, though rather keep it nicely separated. One very concrete difference this creates ATM is that the tests can run without having PySide installed, as per #23.

You are right about the name, I'm also not happy about it. Will try kicking this into a bit better form.

Eximius commented:
Okay.

So, I assume the goal of this was to add a youtube blacklist configuration and you went on a tangent at the start to reformat the command handling, which, admittedly, left a bit to be desired, such as registering new commands seamlessly to the system as you tried to accomplish with your 'polymorphic' command classes.

However, what you made only built more complexity while achieving nothing more than, potential additions could have included argument type enforcing, input data sanitization in such a complex system.

A MessageRouter could totally work, while is a bit more complex. Considering that we have to deal with per-command time limiters it might be a good idea.

AAaaaanyway, your class encapsulation of commands, while works, is waaay to overly complex for simple message handling. In python you might as well use simple functions to handle the messages and keep a list of handling functions in the 'router'.

Apart from that, the actual problems:
1. The blacklisting has a big flaw where it could return less than five (5) casts quite a lot of the time.
2. You removed hitbox stream listing. (Which is a regression)
3. You moved the twitch api encapsulation out into a message handler, which makes it less maintainable. (Which is a ~regression)

Overall, you expanded 95 lines of code to 195 lines of code (not including tests, bet system), without really doing much and overall regressing. I would not merge this.

JeroenDeDauw commented:
Thanks for your feedback @Eximius 

> So, I assume the goal of this was to add a youtube blacklist configuration and you went on a tangent at the start

Incorrect. Please look at my original message and what it links.

> However, what you made only built more complexity while achieving nothing more than, potential additions could have included argument type enforcing, input data sanitization in such a complex system.

I very much disagree with this. Complexity does not arise from the number of containers, it arises from the amount of stuff put together. The average NPath complexity went way down, and so did the Weighted Class Complexity. Many Single Responsibility Principle violations where solves and the code was made testable. The goal of these things is to make the code more maintainable, easier to extend, and easier to understand. If this sounds crazy to you, I suggest you go read Clean Code by Robert C Martin, well worth the time.

Type enforcing and input sanitation are not goals I pursued here.

> In python you might as well use simple functions to handle the messages and keep a list of handling functions in the 'router'.

Notice how there is no type for message handlers. The router just deals with functions, and indeed has a lambda for one command. One can on a command basis make the judgement on how to best organize the code. And in case of the casts and streams commands, I judged that a class was in order. Looking at the result, this seems to be quite obvious now. Both classes contain fields and multiple methods, with a good level of cohesion.

> The blacklisting has a big flaw where it could return less than five (5) casts quite a lot of the time.

As noted, I am well aware of this. To me it seems like a more complex approach is not needed.

> You removed hitbox stream listing. (Which is a regression)

When I tried writing a test for this functionality, I found that it was broken, so I removed it. If I'm wrong, and it currently works, please demonstrate so.

> You moved the twitch api encapsulation out into a message handler, which makes it less maintainable. (Which is a ~regression)

Are you talking about the `twitch.py` file which I removed? If you think that was well encapsulated and better from a maintenance perspective than the current code I again recommend you go read Clean Code or similar books.

JeroenDeDauw commented:
Continuing with #22
